operator/src/main/java/oracle/kubernetes/operator/utils/DomainUpgradeUtils.java

```diff
@@ -0,0 +1,386 @@
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_DEFAULT_INIT_CONTAINER_COMMAND;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_INIT_CONTAINER_NAME_PREFIX;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_INIT_CONTAINER_WRAPPER_SCRIPT;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_TARGET_PATH;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_VOLUME_NAME_PREFIX;
+
+@SuppressWarnings({"Convert2MethodRef", "unchecked", "rawtypes"})
+public class DomainUpgradeUtils {
+  private static final LoggingFacade LOGGER = LoggingFactory.getLogger("Webhook", "Operator");
+
+  public static final String API_VERSION_V9 = "weblogic.oracle/v9";
+  public static final String API_VERSION_V8 = "weblogic.oracle/v8";
+  private volatile int containerIndex = 0;
+
+  public ConversionReviewModel readConversionReview(String resourceName) {
+    return getGsonBuilder().fromJson(resourceName, ConversionReviewModel.class);
+  }
+
+  /**
+   * Create the conversion review response.
+   * @param conversionRequest The request to be converted.
+   * @return ConversionResponse The response to the conversion request.
+   */
+  public ConversionResponse createConversionResponse(ConversionRequest conversionRequest) {
+    List<Object> convertedDomains = new ArrayList<>();
+
+    conversionRequest.getObjects()
+            .forEach(domain -> convertedDomains.add(
+                    convertDomain((Map<String,Object>) domain, conversionRequest.getDesiredAPIVersion())));
+
+    return new ConversionResponse()
+            .uid(conversionRequest.getUid())
+            .result(new Result().status("Success"))
+            .convertedObjects(convertedDomains);
+  }
+
+  public String writeConversionReview(ConversionReviewModel conversionReviewModel) {
+    return getGsonBuilder().toJson(conversionReviewModel, ConversionReviewModel.class);
+  }
+
+  /**
+   * Convert the domain to desired API version.
+   * @param domain Domain to be converted.
+   * @return Domain The converted domain.
+   */
+  public Object convertDomain(Map<String, Object> domain, String desiredAPIVersion) {
+    Map<String, Object> spec = getSpec(domain);
+    if (spec == null) {
+      return domain;
+    }
+    LOGGER.fine("Converting domain " + domain + " to " + desiredAPIVersion + " apiVersion.");
+
+    String apiVersion = (String)domain.get("apiVersion");
+    Map<String, Object> adminServerSpec = (Map<String, Object>) spec.get("adminServer");
+    Boolean adminChannelPortForwardingEnabled = (Boolean)Optional.ofNullable(adminServerSpec)
+            .map(as -> as.get("adminChannelPortForwardingEnabled")).orElse(null);
+    if ((adminChannelPortForwardingEnabled == null) && (API_VERSION_V8.equals(apiVersion))) {
+      Optional.ofNullable(adminServerSpec).map(as -> as.put("adminChannelPortForwardingEnabled", false));
```

Comment:
Why does new code for converting aux volumes need to handle 'admin channel port forwarding enabled'?

Reply:
This is for disabling the port forwarding by default in 4.0 if it was not explicitly enabled in 3.x. This is to prevent pods from rolling during operator upgrade scenario.

Reply:
ok